File: Assets/Scripts/CaseMaroon/WorldMap/NoiseGenerator.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;

namespace CaseMaroon.WorldMap
{
    public class NoiseGenerator : MonoBehaviour
    {
        [Header("Landscape Noise")]

        public NoiseSettings landNoiseSettings;

        [Header("Rain Noise")]

        public NoiseSettings rainNoiseSettings; 

        [Header("Temperature Noise")]

        public NoiseSettings tempNoiseSettings;

        float[,] landValues;
        float[,] rainValues;
        float[,] tempValues;

        public float NoiseHash { get; private set; }

        AverageValues avg;
        public bool NoiseModified;

        public bool printAverage = false;

        private void OnValidate()
        {
            // everytime a value is changed, the noise is modified, thus we need to recompute the noise
            NoiseModified = true;
        }

        public void ComputeNoises(Vector2Int planetSize, bool multiThread = false)
        {
            // if the noise settings are not modified, then no need to recompute the noise
            if (!NoiseModified && planetSize == landNoiseSettings.gridSize)
            {
                return;
            }

            landNoiseSettings.gridSize = planetSize;
            rainNoiseSettings.gridSize = planetSize;
            tempNoiseSettings.gridSize = planetSize;

            landNoiseSettings.Init();
            rainNoiseSettings.Init();
            tempNoiseSettings.Init();

            NoiseHash = 0;

            avg = new AverageValues(planetSize.x * planetSize.y);
            avg.Init("Land");
            avg.Init("Rain");
            avg.Init("Temp");

            landValues = new float[planetSize.x, planetSize.y];
            rainValues = new float[planetSize.x, planetSize.y];
            tempValues = new float[planetSize.x, planetSize.y];

            if (multiThread)
            {
                Parallel.For(0, planetSize.x, i =>
                {
                    for (int j = 0; j < planetSize.y; j++)
                    {
                        ComputeLandNoise(i, j);
                        ComputeRainNoise(i, j);
                        ComputeTempNoise(i, j);
                    }
                });
            }
            else
            {
                for (int i = 0; i < planetSize.x; i++)
                {
                    for (int j = 0; j < planetSize.y; j++)
                    {
                        ComputeLandNoise(i, j);
                        ComputeRainNoise(i, j);
                        ComputeTempNoise(i, j);

                        NoiseHash += landValues[i, j] + rainValues[i, j] + tempValues[i, j];
                    }
                }
            }



            NoiseModified = false;

            if (printAverage)
            {
                Debug.Log(avg.GetString());
            }
        }

        private void ComputeLandNoise(int x, int y)
        {
            landValues[x, y] = landNoiseSettings.GetNoise(x, y);

            avg.Add("Land", landValues[x, y]);
        }
        private void ComputeRainNoise(int x, int y)
        {
            rainValues[x, y] = rainNoiseSettings.GetNoise(x, y);
            avg.Add("Rain", rainValues[x, y]);
        }
        private void ComputeTempNoise(int x, int y)
        {
            tempValues[x, y] = tempNoiseSettings.GetNoise(x, y);
            avg.Add("Temp", tempValues[x, y]);
        }
        public float GetLandNoise(int x, int y)
        {
            return landValues[x, y];
        }
        public float GetRainNoise(int x, int y)
        {
            return rainValues[x, y];
        }

        public float GetTempNoise(int x, int y)
        {
            return tempValues[x, y];
        }

        public string GetPercentString()
        {
            return avg.GetString();
        }


        [Serializable]
        public struct NoiseSettings
        {
            FastNoiseLite noiseGenerator;
            public FastNoiseLite.FractalType fractalType;
            public FastNoiseLite.NoiseType noiseType;

            public int seed;
            public float frequency;
            public float multiplier;
            public float scale;

            [Range(0, 20)]
            [SerializeField] int fractal;

            [Range(0, 1)]
            [SerializeField] float minValue;

            [Range(0, 1)]
            [SerializeField] float maxValue;

            public Vector2Int offset;

            [NonSerialized]
            [HideInInspector]
            public Vector2Int gridSize;
            // code this part

            public void Init()
            {
                noiseGenerator = new FastNoiseLite(seed);
                noiseGenerator.SetFractalType(fractalType);
                noiseGenerator.SetNoiseType(noiseType);
                noiseGenerator.SetFrequency(frequency);
            }
            public float GetNoise(int x, int y)
            {
                float nx = (float) (x / ((float)gridSize.x + offset.x));
                float ny = (float) (y / ((float)gridSize.y + offset.y));

                float tempNoise = noiseGenerator.GetNoise(nx, ny);

                tempNoise *= multiplier;

                tempNoise = Mathf.Clamp(tempNoise, minValue, maxValue);

                tempNoise = (float)Math.Round(tempNoise, 3);

                return tempNoise;
            }

            public float GetNoise2(int x, int y)
            {
                float tempNoise = noiseGenerator.GetNoise((float)x / (gridSize.x + offset.x), (float)y / (gridSize.y + offset.y));

                tempNoise *= multiplier;

                tempNoise = Mathf.Clamp(tempNoise, minValue, maxValue);

                return tempNoise;
            }
        }

        public struct AverageValues
        {
            public Dictionary<string, List<int>> percents;

            public int max;
            public AverageValues(int max)
            {
                percents = new Dictionary<string, List<int>>();

                this.max = max;
            }
            public void Add(string k, float num)
            {
                int index = (int)(num * 10);

                index = Mathf.Clamp(index, 0, 9);

                percents[k][index]++;
            }

            public void Init(string k)
            {
                percents.Add(k, new List<int>());

                for (int i = 0; i < 10; i++)
                {
                    percents[k].Add(0);
                }
            }

            public string GetString()
            {
                string str = "";

                foreach (string key in percents.Keys)
                {
                    str += key + ": ";

                    foreach (int i in percents[key])
                    {
                        float p = (float)i / max;

                        str += p.ToString("F") + " --- ";
                    }

                    str += "\n";
                }

                return str;
            }
        }
    }
}

File: Case Maroon_Backend/src/models/WorldMap/NoiseGenerator.js
import FastNoiseLite from "fastnoise-lite";
import { Vector2 } from "js-vectors";

export class NoiseGenerator {
  constructor(gridSize, noiseConfig) {
    this.gridSize =
      gridSize instanceof Vector2
        ? gridSize
        : new Vector2(gridSize.x, gridSize.y);

    this.landNoiseSettings =
      noiseConfig.landNoiseSettings instanceof NoiseSettings
        ? noiseConfig.landNoiseSettings
        : new NoiseSettings(noiseConfig.landNoiseSettings);

    // fractype typ and opther enums still have to be converted to the correct type

    this.rainNoiseSettings =
      noiseConfig.rainNoiseSettings instanceof NoiseSettings
        ? noiseConfig.rainNoiseSettings
        : new NoiseSettings(noiseConfig.rainNoiseSettings);

    this.tempNoiseSettings =
      noiseConfig.tempNoiseSettings instanceof NoiseSettings
        ? noiseConfig.tempNoiseSettings
        : new NoiseSettings(noiseConfig.tempNoiseSettings);

    // console.log("Land Noise Settings:", this.landNoiseSettings);
    // console.log("Rain Noise Settings:", this.rainNoiseSettings);
    // console.log("Temp Noise Settings:", this.tempNoiseSettings);

    this.noiseHash = 0;
  }

  ComputeNoise() {
    const width = this.gridSize.x;
    const height = this.gridSize.y;

    this.landNoiseSettings.init(this.gridSize);
    this.rainNoiseSettings.init(this.gridSize);
    this.tempNoiseSettings.init(this.gridSize);

    this.landValues = new Array(width);
    this.rainValues = new Array(width);
    this.tempValues = new Array(width);

    for (let x = 0; x < width; x++) {
      this.landValues[x] = new Array(height);
      this.rainValues[x] = new Array(height);
      this.tempValues[x] = new Array(height);

      for (let y = 0; y < height; y++) {
        const landVal = this.landNoiseSettings.getNoise(x, y, this.gridSize);
        const rainVal = this.rainNoiseSettings.getNoise(x, y, this.gridSize);
        const tempVal = this.tempNoiseSettings.getNoise(x, y, this.gridSize);

        this.landValues[x][y] = landVal;
        this.rainValues[x][y] = rainVal;
        this.tempValues[x][y] = tempVal;

        // Simple sum hash, could be improved
        this.noiseHash += landVal + rainVal + tempVal;
      }
    }
  }

  getNoiseHash() {
    return this.noiseHash;
  }

  getLandNoise(x, y) {
    return this.landValues[x][y];
  }

  getRainNoise(x, y) {
    return this.rainValues[x][y];
  }

  getTempNoise(x, y) {
    return this.tempValues[x][y];
  }
}

export class NoiseSettings {
  constructor({
    fractalType,
    noiseType,
    seed,
    frequency,
    multiplier,
    scale,
    fractal,
    minValue,
    maxValue,
    offset,
  }) {
    this.fractalType = fractalType;
    this.noiseType = noiseType;
    this.seed = seed;
    this.frequency = frequency;
    this.multiplier = multiplier;
    this.multiplier = Math.round(multiplier * 100) / 100;
    this.scale = scale;
    this.fractal = fractal;
    this.minValue = minValue;
    this.maxValue = maxValue;
    this.offset = offset; // assuming simple {x,y} object

    this.gridSize = null;
    this.noiseGenerator = null;
  }

  init(gridSize) {
    this.gridSize = gridSize;

    this.noiseGenerator = new FastNoiseLite();

    this.noiseGenerator.SetSeed(this.seed);

    const fractals = Object.values(FastNoiseLite.FractalType)[this.fractalType];

    const noiseTypes = Object.values(FastNoiseLite.NoiseType)[this.noiseType];

    this.noiseGenerator.SetFractalType(fractals);
    this.noiseGenerator.SetNoiseType(noiseTypes);
    this.noiseGenerator.SetFrequency(this.frequency);
  }

  getNoise(x, y, planetSize) {
    const nx = x / (planetSize.x + this.offset.x);
    const ny = y / (planetSize.y + this.offset.y);

    // Step 2: Get raw noise
    let tempNoise = this.noiseGenerator.GetNoise(nx, ny);

    // Step 4: Multiply
    tempNoise *= this.multiplier;

    // Step 5: Clamp
    tempNoise = Math.max(this.minValue, Math.min(this.maxValue, tempNoise));

    // Step 6: Final round
    tempNoise = Math.round(tempNoise * 1e2) / 1e2;

    return tempNoise;
  }

  toJson() {
    return {
      fractalType: this.fractalType,
      noiseType: this.noiseType,
      seed: this.seed,
      frequency: this.frequency,
      multiplier: this.multiplier,
      scale: this.scale,
      fractal: this.fractal,
      minValue: this.minValue,
      maxValue: this.maxValue,
      offset: this.offset,
    };
  }
}

export class NoiseConfig {
  constructor(landNoiseSettings, rainNoiseSettings, tempNoiseSettings) {
    this.landNoiseSettings = landNoiseSettings;
    this.rainNoiseSettings = rainNoiseSettings;
    this.tempNoiseSettings = tempNoiseSettings;
  }
}
